File: python/1-2.py
import base64
import binascii

str1 = '1c0111001f010100061a024b53535009181c'
str2 = '686974207468652062756c6c277320657965'


def hex_to_b64(hex_val):
    if isinstance(hex_val, str):
        hex_val = binascii.unhexlify(hex_val)
        b64_val = base64.b64encode(hex_val)
        return b64_val
    return 'err'


# Takes in two hex strings and xors them
def xor(sx, sy):
    return '{:x}'.format(int(sx, 16) ^ int(sy, 16))


def xor2(sx, sy):
    sx = binascii.unhexlify(sx)
    sy = binascii.unhexlify(sy)
    s = "".join(chr(x ^ y) for x, y in zip(sx, sy))
    return "".join('{:x}'.format(ord(c)) for c in s)


if __name__ == "__main__":
    val = xor(str1, str2)
    print(val)
    val = xor2(str1, str2)
    print(val)

File: rust/1-2.rs
extern crate serialize;

use std::str;

// Convert ascii string to hex string
fn hexlify(str1:String) -> String {
  use serialize::hex::{ToHex};

  return str1.as_slice().as_bytes().to_hex();
}

// Convert hex string to ascii string
fn unhexlify(hex_string:&str) -> String {
  use serialize::hex::{FromHex};

  let byteVec:Vec<u8> = hex_string.from_hex().unwrap();
  let asciiStr = str::from_utf8(byteVec.as_slice());
  println!("{}", asciiStr);
  return String::from_str(asciiStr.unwrap());
}

fn fixed_xor(str1:&str, str2:&str) -> String {
  // Unhexlify strings
  let unhex_str1 = unhexlify(str1);
  let unhex_str2 = unhexlify(str2);

  // Convert the two strings to bytes and zip together.
  let zipped:Vec<(&u8, &u8)> = unhex_str1.as_bytes().iter().zip(
    unhex_str2.as_bytes().iter()
  ).collect();

  // Iterate over byte tuples and xor values
  let mut xored:Vec<u8> = Vec::new();
  for tup in zipped.iter() {
    let x1 = *tup.val0() as u8;
    let x2 = *tup.val1() as u8;
    xored.push(x1 ^ x2);
  }

  // Convert to chars and join into string
  let asciiStr:String = xored.iter().map(|x| *x as char).collect();
  println!("{}", asciiStr);

  // Convert xored answer back into hex string
  return hexlify(asciiStr);
}

fn main() {
  // Strings provided by challenge.
  // The XOR combination of "str1" and "str2" should be "answer"
  let str1 = "1c0111001f010100061a024b53535009181c";
  let str2 = "686974207468652062756c6c277320657965";
  let answer = "746865206b696420646f6e277420706c6179";
  let myAnswer = fixed_xor(str1, str2);

  println!("{}", myAnswer);
  assert!(answer == myAnswer.as_slice());
}
